[Update] cost is added to all implicit APIs

LambdaBen opened the pull request:
describe에 작성된 Implicit API 별로 cost를 추가하였습니다. 

cost 계산법은 아래와 같습니다. 

1. https://docs.upbit.com/kr/docs/user-request-guide를 기준으로 API 별 RPS(Request Per Second)을 확인합니다.
2. 예를 들어, 초당 8회 호출 가능한 API는 125ms 입니다. (1000 ms / 8)
3. 이를 기준으로 125ms = 1000 ms / (rateLimit * cost)를 계산해 적절한 cost를 확인합니다.

LambdaBen commented on ts/src/upbit.ts:
```diff
@@ -103,53 +103,54 @@ export default class upbit extends Exchange {
-                        'orderbook',
-                    ],
+                    'get': {
+                        //  https://docs.upbit.com/kr/docs/user-request-guide
```
Unified API 설명에서 불필요한 주석입니다.

joseph-lambda256 commented on ts/src/upbit.ts:
```diff
@@ -103,53 +103,54 @@ export default class upbit extends Exchange {
-                    ],
+                    'get': {
+                        //  https://docs.upbit.com/kr/docs/user-request-guide
+                        'market/all': 3,
+                        'candles/{timeframe}': 3,
+                        'candles/{timeframe}/{unit}': 3,
+                        'candles/minutes/{unit}': 3,
+                        'candles/minutes/1': 3,
+                        'candles/minutes/3': 3,
+                        'candles/minutes/5': 3,
+                        'candles/minutes/10': 3,
+                        'candles/minutes/15': 3,
+                        'candles/minutes/30': 3,
+                        'candles/minutes/60': 3,
+                        'candles/minutes/240': 3,
+                        'candles/days': 3,
+                        'candles/weeks': 3,
+                        'candles/months': 3,
+                        'trades/ticks': 3,
+                        'ticker': 3,
+                        'orderbook': 3,
```
내부적으로 rateLimit을 위해 token bucket algorithm을 사용하고 있는데요. rateLimit에 따라 토큰이 소숫점 단위로 채워지며 최대 기본 capacity가 1로 되어 있습니다. 따라서 cost를 0.11 정도로 설정해야 1초에 10번 미만으로 실행될 것 같습니다.

참고:
- https://en.wikipedia.org/wiki/Token_bucket

LambdaBen replied:
말씀하신 0.11은 rateLimit이 1000 기준으로 계산하신 것 같은데 혹시 rateLimit을 33으로 설정했을 때 위의 기준이 적합한지 확인해 주시면 감사하겠습니다.

joseph-lambda256 commented on ts/src/upbit.ts:
```diff
@@ -103,53 +103,54 @@ export default class upbit extends Exchange {
-                    'delete': [
-                        'order',
-                    ],
+                    'get': {
+                        'accounts': 1,
+                        'orders/chance': 1,
+                        'order': 1,
+                        'orders': 1,
+                        'orders/closed': 1,
+                        'orders/open': 1,
+                        'orders/uuids': 1,
+                        'withdraws': 1,
+                        'withdraw': 1,
+                        'withdraws/chance': 1,
+                        'deposits': 1,
+                        'deposit': 1,
+                        'deposits/coin_addresses': 1,
+                        'deposits/coin_address': 1,
+                    },
+                    'post': {
+                        'orders': 4,
+                        'withdraws/coin': 1,
+                        'withdraws/krw': 1,
+                        'deposits/generate_coin_address': 1,
+                    },
+                    'delete': {
+                        'order': 1,
+                    },
```
내부적으로 rateLimit을 위해 token bucket algorithm을 사용하고 있는데요. rateLimit에 따라 토큰이 소숫점 단위로 채워지며 최대 기본 capacity가 1로 되어 있습니다. 따라서 cost를 0.035 정도로 설정해야 1초에 30번 미만으로 실행될 것 같습니다.

참고:
- https://en.wikipedia.org/wiki/Token_bucket

LambdaBen replied:
말씀하신 0.035는 rateLimit이 1000 기준으로 계산하신 것 같은데 혹시 rateLimit을 33으로 설정했을 때 위의 기준이 적합한지 확인해 주시면 감사하겠습니다.

joseph-lambda256 commented on ts/src/upbit.ts:
```diff
@@ -103,53 +103,54 @@ export default class upbit extends Exchange {
+                        'deposits/coin_address': 1,
+                    },
+                    'post': {
+                        'orders': 4,
```
내부적으로 rateLimit을 위해 token bucket algorithm을 사용하고 있는데요. rateLimit에 따라 토큰이 소숫점 단위로 채워지며 최대 기본 capacity가 1로 되어 있습니다. 따라서 cost를 0.13 정도로 설정해야 1초에 8번 미만으로 실행될 것 같습니다.

참고:
- https://en.wikipedia.org/wiki/Token_bucket

LambdaBen replied:
말씀하신 0.13은 rateLimit이 1000 기준으로 계산하신 것 같은데 혹시 rateLimit을 33으로 설정했을 때 위의 기준이 적합한지 확인해 주시면 감사하겠습니다.

joseph-lambda256 commented on ts/src/upbit.ts:
```diff
@@ -23,7 +23,7 @@ export default class upbit extends Exchange {
             'countries': [ 'KR' ],
             'version': 'v1',
-            'rateLimit': 1000,
+            'rateLimit': 33,
```
1밀리초 단위로 채워지는 토큰의 수를 결정하는 값으로 `1000`을 유지해야 할 것 같습니다.

LambdaBen replied:
해당 값은 업비트 측의 요구사항으로 연속된 API 호출 사이 최소 33ms 까지는 감당할 수 있다는 답변을 받아 작성한 값 입니다.
따라서 33으로 유지해야 할 것 같습니다.

LambdaBen replied:
전체 변경 완료하였습니다. 확인 부탁드립니다.

joseph-lambda256 reviewed (commented):
스펙상 API별로 별도의 rate limit이 있는 것으로 알고 있습니다. API에 따라 별도의 rate limit이 걸리도록 수정되면 좋을 것 같습니다.

joseph-lambda256 reviewed (dismissed):
LGTM

joseph-lambda256 commented:
@LambdaBen 아래와 같이 설정하면 rateLimit을 벗어나지 않고 사용할 수 있을 것 같습니다.

- rateLimit: 37
- public api: 3.375
- private api: 1
- order api: 4.5

계산 방식은 아래와 같습니다.

- private api 초당 호출 수 = 30 - 3 = 27
- order api 초당 호출 수 = 8 - 2 = 6
- public api 초당 호출 수 = 10 - 2 = 8
- rate limit = 1000 / (private api 초당 호출 수) = 37.037037... ≈ 37
- private api cost
  = (private api 초당 호출 수) / (private api 초당 호출 수)
  = 27 / 27
  = 1
- order api cost
  = (private api 초당 호출 수) / (order api 초당 호출 수)
  = 27 / 6
  = 4.5
- public api cost
  = (private api 초당 호출 수)/ (public api 초당 호출 수)
  = 27 / 8
  = 3.375

기본적으로 2를 뺀 이유는 throttler 로직상 buket에 token이 1이 차있도록 되어 있는 상황에서 cost가 감소되기 때문에 처음 요청에 대해서는 cost가 1이 빠진 값으로 생각해야 합니다. 따라서 일괄적으로 RPS를 1 제외했고요, 추가로 ratelimit을 걸리지 않게 하기 위해 1을 더 뺐습니다. private api의 경우 bucket의 크기와 cost가 동일한데, 로직상 0보다 크거나 같으면 호출을하도록 되어 있어 추가 1을 더 뺐습니다.

LambdaBen commented:
말씀해주신 내용에 따라 cost 변경하였습니다! 확인 감사합니다.

joseph-lambda256 reviewed (approved):
LGTM

baileySoobinPark commented on ts/src/upbit.ts:
```diff
@@ -103,53 +103,54 @@ export default class upbit extends Exchange {
             },
             'api': {
                 'public': {
-                    'get': [
```
기존에 Implicit API 목록을 관리하는 객체의 포맷을 변경하고 새로운 값을 추가한 부분이므로,
//'endpoint','API Cost' 등의 형태로 해당 숫자가 API cost임을 직관적으로 알 수있도록 주석 추가했으면 좋겠습니다.

baileySoobinPark commented on ts/src/upbit.ts:
```diff
@@ -103,53 +103,54 @@ export default class upbit extends Exchange {
-                    ],
+                    'get': {
+                        //  https://docs.upbit.com/kr/docs/user-request-guide
+                        'market/all': 3.375,
```
3.375라는 숫자가 직관적으로 이해되지 않는 숫자일 수 있으므로, 
rateLimit과 Cost 파트에 대해 간단한 주석 추가했으면 좋겠습니다.
37이라는 rateLimit설정값과 3.375라는 숫자가 결국 8RPS를 설정한다는 점을 표현해주세요.

LambdaBen commented:
Implicit API 전부 추가할 때 까지 해당 건 PR은 홀드하겠습니다.

LambdaBen committed:
feat(upbit): Add cost to implicit API and comment.

1. Changed the rateLimit. 1000 -> 50
2. Added cost to all implicit APIs based on the rateLimit and the request limits described in the exchange documentation.
    - Docs Link: https://docs.upbit.com/kr/docs/user-request-guide

Private APIs
	1.	8 RPS
	- POST /orders
	- POST /orders_cancel_and_new
	2.	0.5 RPS
	- DELETE /orders/open
	3.	30 RPS
	- All other private endpoints

Public APIs
	1.	10 RPS
	- All public endpoints

LambdaBen commented:
@baileySoobinPark @joseph-lambda256 
안녕하세요. 본 PR 코드를 다음과 같이 수정하였습니다.

1. rateLimit 50
- 초당 안정적인 요청 수 조정을 위해 33과 같은 숫자가 아닌 50으로 설정하였습니다.

2. rateLimit, RPS 기반으로 cost를 다시 수정하였습니다.
Private APIs
	1.	8 RPS
	- POST /orders
	- POST /orders_cancel_and_new
	2.	0.5 RPS
	- DELETE /orders/open
	3.	30 RPS
	- All other private endpoints

Public APIs
	1.	10 RPS
	- All public endpoints

3. 주석으로 array => object 변경 명시 및 계산 방법, travelRule 검증 시, 동일 주소 10분에 1회 내용 추가 하였습니다.

baileySoobinPark reviewed (approved):
LGTM
